### DISCUSSION TITLE
Is there a way to mock a sdbus::IConnection?

### DISCUSSION THREAD

**Discussion by dakhouya** [Kistler-Group/sdbus-cpp#163] [Q&A] [answered]
Hi all,

I'm searching for a way to mock completely a sdbus::IConnection to be able to test an object that is extended from an adaptor without having a real dbus connection? 

I did a mock class like this one:

```
class MockSdbusConnection : public sdbus::IConnection {
public:
    MockSdbusConnection() = default;
    ~MockSdbusConnection() override = default;
    MOCK_METHOD1(requestName, void(const std::string& name));
    MOCK_METHOD1(releaseName, void(const std::string& name));
    MOCK_CONST_METHOD0(getUniqueName, std::string());
    MOCK_METHOD0(enterEventLoop, void());
    MOCK_METHOD0(enterEventLoopAsync, void());
    MOCK_METHOD0(leaveEventLoop, void());
    MOCK_METHOD1(addObjectManager, void(const std::string& objectPath));
    MOCK_CONST_METHOD0(getEventLoopPollData, PollData());
    MOCK_METHOD0(processPendingRequest, bool());
    MOCK_METHOD1(setMethodCallTimeout, void(uint64_t timeout));
    MOCK_CONST_METHOD0(getMethodCallTimeout, uint64_t());
};
```

Without providing any EXPECT_CALL from this mock object I'm getting "[org.freedesktop.DBus.Error.InvalidArgs] Connection is not a real sdbus-c++ connection (Invalid argument)" since it's a mock connection. Would there be an other strategy to achieve what I am searching for?

**Comment by sangelovic** [comment, marked as answer]
@dakhouya No, there is no way to do that. `Connection`s have internal base classes. The adaptor/proxy classes are coupled to real sdbus-c++ connection, by design. However, you can decouple the server logic from your object class and test it in unit tests independently from sdbus-c++. A few approaches come to my mind:

* Simply, turn your object class into a template class (`template<class Base> class MyObject : public Base {...};`, where in production you'd instantiate it with real `sdbus::AdaptorInterfaces` template class as Base template argument, in tests you'd provide a dummy class as Base template argument. In this case you can mock `sdbus::IConnection` and push it into `MyObject` constructor.
* Or extract the logic completely away from `MyObject` into a separate class, and `MyObject` would just decorate that class' instance...

Additionally, on top of that, you can test the integration of sdbus-c++ classes, your object class and decoupled logic in higher-level integration tests.

**Reply by dakhouya** [reply]
Thanks for the answer! I decided wrote integration test instead of unit tests for this specific case.
